### ISSUE TITLE
is there a way to replace all of the global functions with an object? 

### ISSUE THREAD

**Issue by mx2323** [laverdet/isolated-vm#211]
hey there, im a little confused on the syntax for how to make functions available to the isolate context. 

if i have a context object that has stuff like this
```
{
     log: (...args) => {console.log(...args)},
     foo : 42; 
     bar: 'some string'
}
```

how do i pass it into eval() such that code like this would work? 

```
log(foo) //should print "42"
```

i have code that looks like this. its obviously wrong - but im having trouble understand the API & docs and how to achieve this. 
```
    const c = {bar:'some string', foo:42, log(...args) => { console.log(...args)} }
    const isolate = new ivm.Isolate({ memoryLimit: 8 });
    const context = await isolate.createContext()

    const result = await context.evalClosure(
        `
        with ($0){
            return foo
        }`
        , [c], { arguments: { reference: true } })
    return result;
```

**Comment by laverdet** [comment]
Take a look at the documentation. You can't just pass functions between isolates since they're not really "data". It's like asking how you would transmit that `c` object from nodejs on the server to Chrome in the browser.

**Comment by mx2323** [comment]
ive been looking at the docs for awhile.. it looks like you were able to pass the logCallback() object to the isolate in the example....?

**Comment by mx2323** [comment]
hey @laverdet 

i put together a sandbox here. https://codesandbox.io/s/hungry-currying-o783s?file=/src/index.js

would you mind taking a look? it appears im really close! - but im getting 
```
TypeError: A non-transferable value was passed
    at global.registerMaterialization (<isolated-vm>:2:12)
    at <isolated-vm>:1:1
    at (<isolated-vm boundary>)
    at /sandbox/src/index.js:64:9
```

at line 64.... 

when i replace the 2nd parameter {} with "1", it appears to start working... so something about that {} is causing 'A non-transferable value was passed'...?

i tried looking through the issues and the docs, but i'm still unclear about what this might mean?

**Comment by laverdet** [comment]
You have `{ copy: false }` there for some reason. It needs to be true.

**Comment by mx2323** [comment]
@laverdet NICE! THANKS so much for putting this library together this is amazing. 

i changed the copy:false to copy:true on line 47. now i can register an empty '{}'. that is so cool. 

OK - one more ask~

would it be possible instead of passing {} as the 2nd parameter to pass something like {'func': () => {console.log("foo")} there? 

that would... mean, that im passing a function inside of the isolate to stuff outside of it to call... sounds kind of crazy.

**Comment by laverdet** [comment]
You have to do them individually. If you want to pass an object with a bunch of functions you can do something like this:

```js
const a = {
  foo: new ivm.Reference(() => {}),
};
```

They will end up as references to functions which need to be called in a special way. This is covered in the documentation.

**Comment by mx2323** [comment]
ok.. i wish that i could do that. ive honestly read the documentation multiple times now and i still dont know understand. 
i would love to help with writing documentation or examples...

**Comment by mx2323** [comment]
hey marcel, 

im really sorry for not understanding... how do i pass the ivm object to the context? i didn't see this in the documentation. 

i updated the codesandbox with my attempt - https://codesandbox.io/s/hungry-currying-o783s?file=/src/index.js 
```
//////////// make ivm() function available in context
context.evalClosureSync(`global.ivm = $0`, [ivm], {
  arguments: { reference: true }
});
````

now it says 
```
ivm.Reference is not a constructor
```

again - im really sorry for not understanding this better.. but i'm finding the documentation hard to understand. i would love to submit a PR for the documentation once i figure things out though. 

thanks again.

**Comment by mx2323** [comment]
ok, i just edited the codesandbox so that registerMaterialization is a function now **inside** of the isolate. and im just trying to get the global.handlers that it sets in the sandbox using 

```
const fnReference = context.global.getSync("handlers");
```

but when i call it using 

```
fnReference.apply(undefined, []);
```

 i get 
```
fnReference.apply(undefined, []);
```

**Comment by mx2323** [comment]
updated the codesandbox - https://codesandbox.io/s/hungry-currying-o783s?file=/src/index.js:2208-2262


ok i think i was able to get ivm into the isolate using this code
```
context.global.setSync("ivm", ivm);
```

but when i try to use the code from the example from the docs 
```
fnReference.applySync(undefined, [], { copy: true });
```

i get 
```
TypeError: fnReference.applySync is not a function
```

thoughts?

**Comment by mx2323** [comment]
yowza, was able to get a bit further @laverdet. thanks!

https://codesandbox.io/s/hungry-currying-o783s?file=/src/index.js

theres one more level of crazyness that i need to get to... passing in an object full of functions to the handler function. will work on that a bit more! thanks for your help.

**Comment by mx2323** [comment]
ok i think i understand this a bit better... let me see if i can explain it. 

if a piece of data is a complex object, it needs to be explicit about how it is passed back and forth between isolates and the main node process. with functions, integers, strings, its simple to copy things back and forth...

you can use the ivm.Reference() function to wrap anything so that it can then be passed by reference between the main node js process and an isolate context. you can pass ivm into the isolate context via 
```
context.global.setSync("ivm", ivm);
```

then you can wrap things with ivm.Reference either in the main node process or any isolate context such that they can then be used in other isolate contexts.

if you are trying to call a function inside of an isolate from the main node context, you also need to wrap the parameter you are passing in with ivm.Reference as well.  


so basically,  anytime that you want to move complex data between isolates and the main node process, it needs to be wrapped with ivm.Reference. some things can be copyable, in which case you would use the copy:true option. 

is that about right? 

-----
separately - 

if you have a function that is an ivm.Reference you use the following code to execute it 
```
func.apply(undefined, [args])

**Comment by laverdet** [comment]
You can avoid manually using `Reference` in most cases by using `{reference: true}` in the transfer options.

**Comment by mx2323** [comment]
Question : on line 68 is there a way to make it such that users don’t have to use ivm.reference?

Is there some code that can wrap function call parameter automatically with ivm.Reference? 

https://codesandbox.io/s/hungry-currying-o783s?file=/src/index.js

**Comment by laverdet** [comment]
That's what `{ reference: true }` is for.

**Comment by mx2323** [comment]
In that specific case, im calling a function registerMaterialization() which is in the main node process from some code running in the isolate. so i dont see where i would put {reference:true}? 

it wouldnt go on line 74, because that's for the eval() of the code.... 

and it isn't on line 54 (where i bring registerMaterialization into the isolate context).. it already has reference:{true}, anyways.

**Comment by laverdet** [comment]
It goes on line 49 where you have `{copy:true}`. You've asked for a copy but what you want is a reference.

**Comment by mx2323** [comment]
ahh yeah i just realized that. awesome! 

would you be open to accepting a PR around the documentation? i would be happy to work on it if you are open to it.

**Comment by mx2323** [comment]
Are there any security implications of using reference: true? 

Is the idea that the reference object is a proxy that prevents escaping attacks, etc?
